diff -ru trafficserver-6.1.1.orig/proxy/http/HttpConfig.h trafficserver-6.1.1/proxy/http/HttpConfig.h
--- trafficserver-6.1.1.orig/proxy/http/HttpConfig.h	2016-02-03 11:35:34.000000000 +0900
+++ trafficserver-6.1.1/proxy/http/HttpConfig.h	2016-05-20 17:35:43.263273527 +0900
@@ -613,7 +613,10 @@
   enum {
     CACHE_REQUIRED_HEADERS_NONE = 0,
     CACHE_REQUIRED_HEADERS_AT_LEAST_LAST_MODIFIED = 1,
-    CACHE_REQUIRED_HEADERS_CACHE_CONTROL = 2
+    CACHE_REQUIRED_HEADERS_CACHE_CONTROL = 2,
+    // NOTE: This is a value added by my private patch, so choose a value which is unlikely
+    // to added in the traffciserver upstream.
+    CACHE_REQUIRED_HEADERS_CACHE_CONTROL_S_MAXAGE = 99
   };
 
   enum {
diff -ru trafficserver-6.1.1.orig/proxy/http/HttpTransact.cc trafficserver-6.1.1/proxy/http/HttpTransact.cc
--- trafficserver-6.1.1.orig/proxy/http/HttpTransact.cc	2016-02-03 11:35:34.000000000 +0900
+++ trafficserver-6.1.1/proxy/http/HttpTransact.cc	2016-05-20 17:38:03.573278119 +0900
@@ -6190,6 +6190,14 @@
         }
         break;
 
+      case HttpConfigParams::CACHE_REQUIRED_HEADERS_CACHE_CONTROL_S_MAXAGE:
+        if (!(response->get_cooked_cc_mask() & MIME_COOKED_MASK_CC_S_MAXAGE)) {
+          DebugTxn("http_trans", "[is_response_cacheable] "
+                                 "s-maxage is required");
+          return false;
+        }
+        break;
+
       default:
         break;
       }
