diff -ru trafficserver-6.2.0.orig/proxy/http/HttpTransact.cc trafficserver-6.2.0/proxy/http/HttpTransact.cc
--- trafficserver-6.2.0.orig/proxy/http/HttpTransact.cc	2016-09-30 07:41:54.112970448 +0000
+++ trafficserver-6.2.0/proxy/http/HttpTransact.cc	2016-09-30 07:44:04.037583305 +0000
@@ -5918,8 +5918,9 @@
   //   "must-revalidate header"
   // How about "s-maxage" and "no-cache" directives?
   uint32_t cc_mask;
+  // NOTE: We would like to serve stale cache even if the origin server response has "s-maxage" header.
   cc_mask = (MIME_COOKED_MASK_CC_MUST_REVALIDATE | MIME_COOKED_MASK_CC_PROXY_REVALIDATE | MIME_COOKED_MASK_CC_NEED_REVALIDATE_ONCE |
-             MIME_COOKED_MASK_CC_NO_CACHE | MIME_COOKED_MASK_CC_NO_STORE | MIME_COOKED_MASK_CC_S_MAXAGE);
+             MIME_COOKED_MASK_CC_NO_CACHE | MIME_COOKED_MASK_CC_NO_STORE /* | MIME_COOKED_MASK_CC_S_MAXAGE */);
   if ((cached_response->get_cooked_cc_mask() & cc_mask) || cached_response->is_pragma_no_cache_set()) {
     DebugTxn("http_trans", "[is_stale_cache_response_returnable] "
                            "document headers prevent serving stale");
